src/tbb/global_control.cpp

```diff
@@ -139,11 +142,31 @@ class alignas(max_nfs_size) lifetime_control : public control_storage {
     }
 };
 
-static allowed_parallelism_control allowed_parallelism_ctl;
-static stack_size_control stack_size_ctl;
-static terminate_on_exception_control terminate_on_exception_ctl;
-static lifetime_control lifetime_ctl;
-static control_storage *controls[] = {&allowed_parallelism_ctl, &stack_size_ctl, &terminate_on_exception_ctl, &lifetime_ctl};
+struct alignas(max_nfs_size) control_storage_instance {
+    allowed_parallelism_control     allowed_parallelism_ctl;
+    stack_size_control              stack_size_ctl;
+    terminate_on_exception_control  terminate_on_exception_ctl;
+    lifetime_control                lifetime_ctl;
+};
+static control_storage_instance* control_storage_instance_ptr = nullptr;
+static control_storage *controls[4] = {nullptr, nullptr, nullptr, nullptr};
```

Comment:
Do we really need an extra variable + extra class? Can we just use pointer to array?

Reply:
It’s not really needed, I was trying to follow @kboyarinov’s instructions but I guess I understood them wrongly. So you meant that each control variable is allocated separately, right?

One argument for the struct is that it only needs one allocation so data is more compact.   With individual allocations you have to hard-code the derived class twice too anyways (construction/destruction). On the other hand, one can argue that the struct is slightly less maintainable, so I’d understand if you still prefer one allocation per class.

Reply:
Nvm, I just realized that every class sits in it’s own cache line so the “compact” argument doesn’t really hold…